pkg/storage/ingest/fetcher.go

```diff
@@ -610,11 +628,48 @@ func casHWM(highWwatermark *atomic.Int64, newHWM int64) {
 	}
 }
 
+type inflightFetchWants struct {
+	// queue is the list of all fetchResult of all inflight fetch operations. Pending results
+	// are ordered in the same order these results should be returned to PollFetches(), so the first one
+	// in the list is the next one that should be returned.
+	queue list.List
+
+	// bytes is the sum of the MaxBytes of all fetchWants that are currently inflight.
+	bytes *atomic.Int64
+}
+
+// nextResult is the channel where we expect a worker will write the result of the next fetch
+// operation. This result is the next result that will be returned to PollFetches(), guaranteeing
+// records ordering.
+func (w *inflightFetchWants) nextResult() chan fetchResult {
+	if w.queue.Len() == 0 {
+		return nil
+	}
+	return w.queue.Front().Value.(fetchWant).result
+}
+
+func (w *inflightFetchWants) count() int {
+	return w.queue.Len()
+}
+
+func (w *inflightFetchWants) add(nextFetch fetchWant) {
+	w.bytes.Add(int64(nextFetch.MaxBytes()))
+	w.queue.PushBack(nextFetch)
+}
+
+func (w *inflightFetchWants) headResultsExhausted() {
```

Comment:
[nit] The `inflightFetchWants` is essentially a list. I would rename this function in terms or list operations (remove the first item) and how the actual action that triggered it (we've done consuming the first item).

```suggestion
func (w *inflightFetchWants) removeNextResult() {
```